File: sdk-javascript/src/recipient/FileLocalKeyProvider.ts
// Copyright (c) HealthFlow Medical HCX. Licensed under Apache 2.0.

import { readFile } from 'node:fs/promises';

import { type KeyLike, importPKCS8 } from 'jose';

import { JWE_ALG } from '../crypto/JweAlgorithms.js';
import { KeyUnavailableError } from '../exceptions/HfcxError.js';
import type { LocalKeyProvider } from './LocalKeyProvider.js';

/**
 * Loads the recipient's RSA private key from a PKCS#8 PEM file on
 * every {@link getPrivateKey} call so rotations take effect without
 * restarting the host process. Sister to Java's
 * `FileLocalKeyProvider`, Python's `FileLocalKeyProvider`, and .NET's
 * `FileLocalKeyProvider`.
 */
export class FileLocalKeyProvider implements LocalKeyProvider {
  constructor(private readonly path: string) {
    if (!path) throw new TypeError('path is required');
  }

  async getPrivateKey(): Promise<KeyLike> {
    let pem: string;
    try {
      pem = await readFile(this.path, 'utf8');
    } catch (err) {
      throw new KeyUnavailableError(
        `Failed to read recipient private key from '${this.path}': ${
          (err as Error).message ?? String(err)
        }`,
      );
    }
    try {
      return await importPKCS8(pem, JWE_ALG);
    } catch (err) {
      throw new KeyUnavailableError(
        `Failed to parse recipient private key PEM at '${this.path}': ${
          (err as Error).message ?? String(err)
        }`,
      );
    }
  }
}

File: sdk-dotnet/src/HealthFlow.Hfcx.Sdk/Recipient/FileLocalKeyProvider.cs
// Copyright (c) HealthFlow Medical HCX. Licensed under Apache 2.0.

using System;
using System.IO;
using System.Security.Cryptography;
using HealthFlow.Hfcx.Sdk.Exceptions;

namespace HealthFlow.Hfcx.Sdk.Recipient;

/// <summary>
/// Loads the recipient's RSA private key from a PKCS#8 PEM file. The
/// key file is re-read on every <see cref="GetPrivateKey"/> call so
/// rotations take effect without restarting the host process. Sister to
/// Java's <c>FileLocalKeyProvider</c> and Python's
/// <c>FileLocalKeyProvider</c>.
/// </summary>
public sealed class FileLocalKeyProvider : ILocalKeyProvider
{
    private readonly string _path;

    /// <summary>Construct over the given filesystem path.</summary>
    public FileLocalKeyProvider(string path)
    {
        ArgumentException.ThrowIfNullOrEmpty(path);
        _path = path;
    }

    /// <inheritdoc />
    public RSA GetPrivateKey()
    {
        string pem;
        try
        {
            pem = File.ReadAllText(_path);
        }
        catch (Exception ex) when (ex is FileNotFoundException
            or DirectoryNotFoundException
            or IOException
            or UnauthorizedAccessException)
        {
            throw new KeyUnavailableException(
                $"Failed to read recipient private key from '{_path}': {ex.Message}");
        }

        var rsa = RSA.Create();
        try
        {
            rsa.ImportFromPem(pem);
            return rsa;
        }
        catch (Exception ex) when (ex is ArgumentException or CryptographicException)
        {
            rsa.Dispose();
            throw new KeyUnavailableException(
                $"Failed to parse recipient private key PEM at '{_path}': {ex.Message}");
        }
    }
}
